overlord/devicestate/handlers_systems.go

```diff
@@ -0,0 +1,307 @@
+}
+
+func (m *DeviceManager) undoCreateRecoverySystem(t *state.Task, _ *tomb.Tomb) error {
+	if release.OnClassic {
+		// TODO: this may need to be lifted in the future
+		return fmt.Errorf("cannot run update gadget assets task on a classic system")
+	}
+
+	st := t.State()
+	st.Lock()
+	defer st.Unlock()
+
+	remodelCtx, err := DeviceCtx(st, t, nil)
+	if err != nil {
+		return err
+	}
+
+	setup, err := taskRecoverySystemSetup(t)
+	if err != nil {
+		return fmt.Errorf("internal error: cannot obtain recovery system setup information")
+	}
+	label := setup.Label
+
+	var undoErr error
+	if err := os.RemoveAll(setup.Directory); err != nil && !os.IsNotExist(err) {
+		t.Logf("when removing recovery system %q: %v", setup.Label, err)
+		undoErr = err
+	} else {
+		t.Logf("removed recovery system directory %v", setup.Directory)
+	}
+	for _, f := range setup.NewCommonFiles {
+		if err := os.Remove(f); err != nil && !os.IsNotExist(err) {
+			t.Logf("when removing seed file %q: %v", f, err)
+			if undoErr == nil {
+				undoErr = err
+			}
+		} else {
+			t.Logf("removed new file %q", f)
+		}
+	}
+
+	if err := boot.DropRecoverySystem(remodelCtx, label); err != nil {
+		return fmt.Errorf("cannot drop a current recovery system %q: %v", label, err)
+	}
+
+	return undoErr
+}
+
+func (m *DeviceManager) doFinalizeTriedRecoverySystem(t *state.Task, _ *tomb.Tomb) error {
+	if release.OnClassic {
+		// TODO: this may need to be lifted in the future
+		return fmt.Errorf("cannot run update gadget assets task on a classic system")
+	}
+
+	st := t.State()
+	st.Lock()
+	defer st.Unlock()
+
+	if ok, _ := st.Restarting(); ok {
+		// don't continue until we are in the restarted snapd
+		t.Logf("Waiting for system reboot...")
+		return &state.Retry{}
+	}
+
+	logger.Noticef("in finalize recovery system")
+
+	remodelCtx, err := DeviceCtx(st, t, nil)
+	if err != nil {
+		return err
+	}
+	isRemodel := remodelCtx.ForRemodeling()
+
+	var triedSystems []string
+	// after rebooting to the recovery system and back, the system got moved
+	// to the tried-systems list in the state
+	if err := st.Get("tried-systems", &triedSystems); err != nil {
+		return fmt.Errorf("cannot obtain tried recovery systems: %v", err)
+	}
+
+	setup, err := taskRecoverySystemSetup(t)
+	if err != nil {
+		return err
+	}
+	label := setup.Label
+
+	// so far so good
+	if !strutil.ListContains(triedSystems, label) {
+		// system failed, trigger undoing of everything we did so far
+		return fmt.Errorf("tried recovery system %q failed", label)
+	}
+
+	if isRemodel {
+		// XXX: candidate system is promoted to the list of good ones once we
+		// complete the whole remodel change
+		logger.Noticef("recovery system will be promoted later")
+		return nil
+	}
+
+	if err := boot.MarkRecoverySystemGood(remodelCtx, label); err != nil {
+		return fmt.Errorf("cannot mark recovery system %q as good: %v", label, err)
+	}
+
+	// tried systems should be a one item list, we can clear it now
+	st.Set("tried-systems", nil)
```

Comment:
we want to mark this task Done here as well

Reply:
Done now